aemcmc/gibbs.py

```diff
@@ -227,6 +229,54 @@ def nbinom_sigmoid_dot_match(
     return X, h, beta_rv
 
 
+def dispersion_term_model(srng):
+    h_rv = srng.gamma(1, 1, size=1)  # a_0 = 1
+    r_rv = srng.gamma(100, 1 / h_rv, size=1)  # b_0 = 1
+    return r_rv
+
+
+def CRT_sum(srng, x, r):
+    max_val = at.max(x)
+    rnd = r / (r + at.arange(max_val + 1))
+    return at.sum(
+        ((srng.uniform(size=(x.shape[0], max_val.eval() + 1)) - rnd) <= 0)[0, :]
+    )
+
+
+def dispersion_term_step(
+    srng: RandomStream,
+    r: TensorVariable,
+    phi: TensorVariable,
+    w: TensorVariable,
+    y: TensorVariable,
+    xb: TensorVariable,
+    T: int,
+) -> Tuple[TensorVariable, TensorVariable]:
+    PsiSig = 1.0 / (phi + w)
+    PsiMu = PsiSig @ ((y - r) / 2 + phi * (xb))
+    rng = np.random.RandomState(2313)
+    Psi = PsiMu + rng.randint(T) * at.sqrt(PsiSig)
+    psi_beta = (Psi - xb) / Psi
+    phi = srng.gamma(1e-2 + T / 2, 1 / (1e-2 + at.sum(psi_beta * psi_beta) / 2), size=1)
```

Comment:
These references to a $\psi$ term lead me to believe that these steps aren't specifically for the $r$ parameter, but other parts of the model that are already covered by the existing PG-based negative-binomial sampler.

That could very easily explain any issues you've seen with this implementation.